### ISSUE TITLE
opening zarr dataset doesn't return as expected

### ISSUE THREAD

**Issue by artttt** [corteva/rioxarray#521]
#### Problem description

I've been experimenting with the gdal zarr driver and `rioxarray.open_rasterio` doesn't return as expected.

Rather then getting the expected DataArray the output is a list of 2 xarray.Datasets that represent the x and y dimensions.

Maybe im missing something but i cant see a way to get the actual data.

However if I remove this section of code it works fine.
https://github.com/corteva/rioxarray/blob/dc8efe36d9e72c93bd20be759b5e464190ba238f/rioxarray/_io.py#L854-L868

This should reproduce the issue.
```python

from affine import Affine
import numpy as np
import rioxarray
import rasterio

o = np.ones((4501, 3039),np.float32)

w_settings = {'driver': 'Zarr',
              'dtype': 'float32',
              'nodata': -9999,
              'width': 3039,
              'height': 4501,
              'count': 1,
              'crs': 3857,
              'transform': Affine(25.0, 0.0, 16556975.0,0.0, -25.0, -4179000.0),
              'blockxsize': 5760,
              'blockysize': 5760,
              'tiled': True}
zarr_settings = dict(
                COMPRESS='BLOSC',
                BLOCKSIZE='5760,5760',
                BLOSC_CNAME="lz4",
                BLOSC_CLEVEL=5,
                BLOSC_SHUFFLE="BYTE",
                #BLOSC_BLOCKSIZE=0,
                BLOSC_NUM_THREADS="ALL_CPUS",
                ARRAY_NAME='data',
                #FORMAT='ZARR_V3',
                )

with rasterio.open(
    'test/test',
    'w',
    **w_settings,
    **zarr_settings
) as dst:
    dst.write(o, 1)
    
da = rioxarray.open_rasterio("test/test")
```

Before you ask, yes i know there are other ways of working with zarrs in xarray but doing it through rioxarray has some advantages such as being able to wrap it in a warpedvrt...which incidentally works ( i guess because gdal ignores the subdatasets when it wraps the zarr in a vrt)
```python
import rasterio
import rioxarray
from rasterio.vrt import WarpedVRT
with rasterio.open("test/test") as src:
    with WarpedVRT(src) as vrt:
        with rasterio.io.MemoryFile() as memfile:
            rasterio.shutil.copy(vrt, memfile.name, driver="VRT")
            new_Vrt = memfile.read().decode("utf-8")

ds_vrt_out = rioxarray.open_rasterio(new_Vrt)
```

#### Environment Information
rioxarray (0.11.1) deps:
  rasterio: 1.2.10
    xarray: 2022.3.0
      GDAL: 3.4.1
      GEOS: None
      PROJ: None
 PROJ DATA: None
 GDAL DATA: None

Other python deps:
     scipy: 1.8.0
    pyproj: 3.3.1

System:
    python: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0]
executable: /env/bin/python
   machine: Linux-5.4.181-99.354.amzn2.x86_64-x86_64-with-glibc2.29

**Comment by artttt** [comment]
Ahh sorry.
The writing from rasterio is working fine (that was just me setting up a test dataset)

The reading in rioxarray is where im seeing an issue


```
da = rioxarray.open_rasterio("test/test")
print(da)
[<xarray.Dataset>
Dimensions:      (band: 1, x: 3039, y: 1)
Coordinates:
  * band         (band) int64 1
  * x            (x) float64 1.656e+07 1.656e+07 ... 1.663e+07 1.663e+07
  * y            (y) float64 1.656e+07
    spatial_ref  int64 0
Data variables:
    X            (band, y, x) float64 ..., <xarray.Dataset>
Dimensions:      (band: 1, x: 4501, y: 1)
Coordinates:
  * band         (band) int64 1
  * x            (x) float64 -4.179e+06 -4.179e+06 ... -4.291e+06 -4.292e+06
  * y            (y) float64 -4.179e+06
    spatial_ref  int64 0
Data variables:
    Y            (band, y, x) float64 ...]

```

**Comment by snowman2** [comment]
> However if I remove this section of code it works fine.

Usually there isn't data in the main raster when subdatasets exist ... I wonder if that is the intended behavior?

**Comment by artttt** [comment]
A simple fix might be to only switch to reading subsets if they exist and there is no data in the main raster.

**Comment by snowman2** [comment]
> A simple fix might be to only switch to reading subsets if they exist and there is no data in the main raster.

This seems to be a driver specific issue. When opening `test/test_data/input/tmmx_20190121.nc`, it has data in the main raster and the subdatasets.
```python
>>> import rasterio
>>> rds = rasterio.open("tmmx_20190121.nc")
>>> rds_sub = rasterio.open('NETCDF:"tmmx_20190121.nc":air_temperature')
>>> assert (rds.read() == rds_sub.read()).all()
>>> assert rds.crs == rds_sub.crs
```

Though, most of the time, there isn't data in the main raster when there are subdatasets.

**Comment by snowman2** [comment]
#522 should hopefully address this specific scenario without impacting other datasets.

**Comment by rouault** [comment]
> Is the behavior expected?

There's no well defined rule on whether a main dataset that has a band (that is not just a place holder for subdatasets) should expose itself in the subdatasets it exposes. This can vary between drivers. But here the driver should probably hide the X and Y variables as it recognizes them to build a geotransform, and they don't really bring any value as being exposed as subdatasets. You can open a ticket in OSGeo/gdal about that

**Comment by artttt** [comment]
That fix seems very specific

This would be problematic for  `test/test_data/input/tmmx_20190121.nc` as well and not solved by that fix?

maybe `open_rasterio` needs to give the user an option to select if they want subdatasets or main datasets.

In contrast to my test setup above im using `open_rasterio` with zarrs that are produced by `xarray.to_zarr`. In those cases the subdatasets may not be called x and y and the crs be recognised by gdal unless the user does a little manipulation of the zarr
like this
```
z = zarr.open(store)
z.data.attrs['_CRS'] = {'wkt':z.spatial_ref.attrs['crs_wkt']}
#and probably rebuild consolidated metadata as well
```
This could be down at zarr creation time but xarray.to_zarr throws an error when writing attributes that are dictionaries as GDAL wants for the CRS.

**Comment by snowman2** [comment]
> This would be problematic for test/test_data/input/tmmx_20190121.nc as well

It appears the same data in the base dataset is the one in the subdataset, so no issues there.

**Comment by snowman2** [comment]
Thanks @rouault :+1:, I opened a ticket: https://github.com/OSGeo/gdal/issues/5681

**Comment by snowman2** [comment]
> There's no well defined rule on whether a main dataset that has a band (that is not just a place holder for subdatasets) should expose itself in the subdatasets it exposes.

Is there a scenario in the GDAL test suite that has a scenario where there is a band in the main dataset with data and has subdatasets?

**Comment by rouault** [comment]
> Is there a scenario in the GDAL test suite where there is a band in the main dataset with data and has subdatasets?

For example, the NITF and GTiff drivers can do that when there are multiple images in a file. GDALOpen() on such files will return a dataset with the first image, and a list of subdatasets listing all images (_SUBDATASET_1 will be the first image)

```
$ gdal_translate autotest/gcore/data/byte.tif test.tif
$ gdal_translate autotest/gcore/data/byte.tif test.tif -co APPEND_SUBDATASET=YES -outsize 200% 200%
$ gdalinfo test.tif
Driver: GTiff/GeoTIFF
Files: test.tif
Size is 20, 20
[...]
Subdatasets:
  SUBDATASET_1_NAME=GTIFF_DIR:1:test.tif
  SUBDATASET_1_DESC=Page 1 (20P x 20L x 1B)
  SUBDATASET_2_NAME=GTIFF_DIR:2:test.tif
  SUBDATASET_2_DESC=Page 2 (40P x 40L x 1B)
```

**Comment by snowman2** [comment]
> and a list of subdatasets listing all images (_SUBDATASET_1 will be the first image)

If I understand correctly, defaulting to only reading subdatasets in this scenario is safe as as `_SUBDATASET_1` will be the same as reading the main dataset.

**Comment by snowman2** [comment]
Reverting the patch as it is best fixed in GDAL: #523

**Comment by artttt** [comment]
Thanks.
